### ISSUE TITLE
proposal: exp/xerrors: add try and handle

### ISSUE THREAD

**Issue by anonld** [golang/go#67913]
### Proposal Details

**Disclaimer: The following was originally posted by u/Enough_Compote697 on r/golang. For greater reach, and especially for the attention from the Go team, re-post here after the deletion of the original post. Please keep in mind that I am _NOT_ the OP of that Reddit thread, so please don’t tag me in this GitHub issue (unless reposting violates the rules here).**

To encourage robust programming by reducing error handling boilerplates, I propose promoting the following functions from a third party library to `exp/xerrors` and eventually making them built-in:

    func handle(h func (error)) {
        r := recover()
        if r == nil {
            return
        }
        if e, ok := r.(error); ok {
            h(e)
            return
        }
        panic(r)
    }

    func check(e error) {
        if e != nil {
            panic(e)
        }
    }

    func try[T any](v T, e error) T {
        check(e)
        return v
    }

    func try2[S, T any](u S, v T, e error) (S, T) {
        check(e)
        return u, v
    }

Technically, a built-in variadic `try()` would eliminate the need of `check`, `try2`, `try3`, etc.

Before commenting/voting on this proposal, please make sure you are familiar with the [Reddiquette](https://support.reddithelp.com/hc/en-us/articles/205926439-Reddiquette), and please abide by it the best you can.

# Examples
---

Ex. 1:

    func main() {
        defer handle(func(e error) { log.Fatal(e) })
        os.Stdout.Close()
        try(fmt.Println(42))
    }

Ex. 2: (see [blog/errors-are-values](https://github.com/golang/website/blob/master/_content/blog/errors-are-values.md))

    func flush(h func(error)) {
        defer handle(h)
        try(fd.Write(p0[a:b]))
        try(fd.Write(p1[c:d]))
        try(fd.Write(p2[e:f]))
    }

Note that with the `try`/`handle` pair, this version of `flush` is much simpler and cleaner than the `errWriter` example in the blog, and much less code is needed.

Ex. 3: (see [issue/32437](https://github.com/golang/proposal/blob/master/design/32437-try-builtin.md))

    func sum(a, b string, h func(error)) int {
        defer handle(h)
        return try(myAtoi(a)) + try(myAtoi(b))
    }

Ex. 4: (see [play/p/bOUz7qx3urD](https://go.dev/play/p/bOUz7qx3urD) and the corresponding r/golang [post](https://www.reddit.com/r/golang/comments/193uk0m/whats_the_best_way_to_handle_error_in_go/))

Ex. 5: (see [design/go2draft-error-handling-overview](https://github.com/golang/proposal/blob/master/design/go2draft-error-handling-overview.md))

    func CopyFile(src, dst string, h func(error)) {
        defer handle(func(err error) { h(fmt.Errorf("copy %s %s: %w", src, dst, err)) }
        r := try(os.Open(src))
        defer r.Close()
        w := try(os.Create(dst))
        defer handle(func(err error) {
            os.Remove(dst)
            panic(err)  // propagate to the handle above for formatting
        })
        defer w.Close() // to close b4 remove, as defer is last-in-fisrt-out
        try(io.Copy(w, r))
    }

Ex. 6: (see [design/go2draft-error-handling](https://github.com/golang/proposal/blob/master/design/go2draft-error-handling.md)

    func SortContents(w io.Writer, files []string, h func(error)) {
        defer handle(func(e error) { h(fmt.Errorf("process: %w", e)) })
        lines := make([][]string, len(files))
        for i, f := range files {
            defer handle(func(e error) { panic(fmt.Errorf("read %s: %w", f, e)) })
            r := try(os.Open(f))
            defer r.Close()
            lines[i] = mySplit(try(io.ReadAll(r)))
        }
        sorted := slices.Sort(slices.Concat(lines))
        for _, line := range sorted {
            try(io.WriteString(w, line))
        }
    }

To prevent mishandling deeply nested error that is out of interest, one may limit the error type to a specific kind, and handle the error only after a type assertion, e.g.

    defer handle(func(e error) {
        if pe, ok := xerrors.As[fs.PathError](e); ok {
            // error handling code
        } else {
            panic(e) // not interested, pretending not intercepted
        }
    })

Ex. 7: (an alternate version of ex. 6, changes marked with ~)

    func SortContents(w io.Writer, files []string~) ~error~ {
        defer handle(func(e error) { ~panic~(fmt.Errorf(...) }
        // ...
        return nil
    }

Since both `return` and `panic` are for propagating error back to the `main()` function, they are nearly identical semantically. That said, if library authors dislike this kind of error propagation, they may just keep using the current idiomatic `if err != nil` to handle errors, or adopt the handler propagation style illustrated above.

In addition, even though this version of `SortContents` panics instead of returning an error, the return argument (`error`) in the function signature can be kept for documentation purpose by treating the signature as a kind of function annotation, but the cost is to add a corresponding trailing `return nil` statement. Although the compilation will fail if this final line is omitted, it seems like a gotcha with stinky code smell.

Ex. 8: (another alternative as in `regexp` and `text/template`)

    func MustSortContents(…) {…}

Note the idiomatic `Must` prefix implies the function panics in case of error. Although it is up to callers to decide whether to ignore or handle the error in case the program panics, personally I am more inclined to proper error handling rather than turning a blind eye on it, as this proposal is all about encouraging robust programming.

# Motivation
---

When there are too many trivial error handling surrounded by boilerplates, their non-trivial counterparts will just nearly disappear, so this proposal tries to promote the visibility of truly important red flags by reducing boilerplates.

Also, by reducing the burdens on programmers, they are less deterred from handle error robustly. (See the [law of effect](https://en.m.wikipedia.org/wiki/Law_of_effect) and [operant conditioning](https://en.m.wikipedia.org/wiki/Operant_conditioning) for more on the topic.)

# Scope
---

This proposal is only about adding 2 functions (`try` and `handle`) to the official repository. Although the introduction of this pair may sparkle debate on error handling style (handler propagation vs error propagation), this proposal is not intended to set a new convention. That is left for the community to discuss and experiment on how to use these 2 functions. If you prefer the idiomatic `return err`, it is fine, and you can just keep using it. This proposal is not aimed at replacing that idiomatic error handling style but just at reducing programmers’ burden by promoting to `try`/`handle` pair from a third party library to the official repository, so please keep in mind that the discussion here should focus on the cost and benefits of such promotion. Thanks for your cooperation.

# (Expected) Benefits
---

- By reducing the burden of programmers, they are less deterred from robust error handling. 
- By refactoring repeating error handling blocks, programmers are writing code less vulnerable to copy-pasting bug, which can be catastrophic. (Below are examples from Google Project Zero)
  1. [Exception-oriented exploitation on iOS](https://googleprojectzero.blogspot.com/2017/04/exception-oriented-exploitation-on-ios.html)
  2. [Oops, I missed it again!](https://googleprojectzero.blogspot.com/2020/11/oops-i-missed-it-again.html)

# (Expected) Cost
---

Basically, when it comes to programming proposal, it is all about trade-off in the following trilemma:
- programmers’ time 
- compilation time 
- execution time 

The latter two can be addressed with technical improvement. For example, `go1.14` reduce the cost of `defer` with [design/34481-opencoded-defers](https://github.com/golang/proposal/blob/master/34481-opencoded-defers.md). Although the benchmark then might now become slightly outdated, the tax of `try`/`handle` still seems far from being prohibitive.

Regarding programmers’ time, it can be further broken into learning time and coding time. Reducing boilerplates means saving coding time in a recurring manner, so the one-time cost of learning looks like a worthy investment, not to mention the potential robustness enhancement.

# Comparisons
---

- vs `design/go2draft-error-handling`: change statements to functions to enhance composablility (see examples above) 
- vs `design/32437-try-builtin`: change implicit `return` to implicit `panic` to reduce surprises, as implicit `panic` already exists while implicit `return` is something new and surprising, violating the principle of least astonishment. 
- same: callers still need to write their error handling logic. 

# FAQ
---

**Q: Why revisit the `try` built-in proposal?**

A: To reduce boilerplates of error handling. It is a pain point of many Go developers according to past surveys.

**Q: Why not the idiomatic `if err != nil { return }` flow but the foreign `throw`-`catch` model?**

A: When most library routines encounter an error, they immediately abort and return a non-`nil` error coupled with 1/2 negligible zero value(s), so replacing `return` with `panic` means nearly semantically the same for the caller.

Moreover, `return err` can coexist with the `panic`-`recover` model, so library authors can choose the error handling flow that fit their needs the most.

**Q: But that violates the Go philosophy. There should be one (and preferably only one) obvious way to do it.**

A: That’s true, but there is nothing new. This proposal is not about adding a 3rd way of error handling to the language, but just about utilising the existing language constructs to reduce boilerplates.

**Q: How about `go try(f(…))`?**

A: Not clear. Before understanding how goroutines would impact `defer`/`panic`/`recover`, it seems prudent to forbid `go try(f(…))` first by
1. issuing a compilation error when try is called in a go statement, and
2. adding a `vet` check for `exp/xerrors` during the transitional stage.

Library authors can keep using the good old idiomatic `if err != nil` check with goroutines as is.

# Discussion
---

1. Examples above simply convert `return err` to `panic`. More complex control flow like `goto`/`break`/`continue` may need extra care when switching the error handling flow to the `panic`-`recover` model. In case simple refactoring is not enough, perhaps just remaining as-is (`if err != nil`) suit such scenario the most. Without real world data, it is hard to tell whether the task of refactoring is easy or daunting. If there are real world examples demonstrating non-trivial refactoring, let’s see if we can write tool to automate the work as in the `rangefunc` [experiment](https://github.com/golang/go/blob/master/src/cmd/compile/internal/rangefunc/rewrite.go). For example, given a function:

      package foo

      //go:generate witherrhandle
      func f(v T) error {
          return g(v)
      }

Running `go generate play.ground/foo` yields:

      func fV2(v T) (err error) {
          fWithHandler(v, func(e error) { err = e })
          return
      }

      func fWithHandler(v T, h func(error)) {
          defer handle(h)
          try(g(v))
      }

2. For the sake of compatibility, current `stdlib` routines would keep panicking or returning errors as is, but the internal may use the new mechanism as long as such changes are invisible to callers. When the timing becomes mature, signatures of newly added functions may form a new idiomatic flow of error handling as shown above. 

3. From a dependency management perspective, promoting `handle` from a third party library to the official repository, meaning establishing a mechanism for panic interception, may do more good than harm, but the pros and cons are not so clear, so it is left for the community to discuss. For example, if package `main` imports package `foo`, and the input sanitizer of `foo.Bar` panics like the `MustXXX` family in package `regexp` and `text/template`, the `main` function, especially that of critical server, would be better-off with panic recovery and graceful error handling.

Alternatives
---

1. The `try` function can be written in a variadic fashion with some cost illustrated below:

      func try(vs ...any) []any {
          if e, ok := vs[len(vs)-1].(error); !ok {
          panic("try: last argument is not error")
      } else if e != nil {
          panic(e)
      }
      return vs[:len(vs)-1]
      }

Although this version of `try` is variadic, it is much less user-friendly. Since it returns `[]any` instead of concrete types, type assertion is needed for devirtualisation. For example, to open a read-only file, one has to write

      f := try(os.Open(name)).(fs.File)

It will be even more clumsy when devirtualising more than one `interface{}`.

2. While `try` is technically most suited to be a built-in function, `handle` is less constrained and it can be a function in the `errors` package instead. Although where will it land eventually is out of scope of this proposal, please feel free to discuss here about the pros and cons of adding one more reserved word. In particular, is `handle` frequently used as a custom variable/function name? If so, it means the built-in `handle` will be often shadowed, rendering it less useful to be a built-in. On the other hand, a nice side-effect of moving the `handle` function to the `errors` package is the possibility to define a visually pleasant `errors.Handler` interface.

Postscript
---
Although this proposal is just about including a pair of trivial functions in the official repository, it is surprisingly hilariously disproportionally lengthy, reminding me the [Parkinson’s law of triviality](https://en.m.wikipedia.org/wiki/Law_of_triviality). The reason is that much length is used for highlighting the differences between the new and the idiomatic error handling flow so as to address the [Fredkin’s paradox](https://en.m.wikipedia.org/wiki/Fredkin%27s_paradox), [psychological inertia](https://en.m.wikipedia.org/wiki/Psychological_inertia) and [path dependence](https://en.m.wikipedia.org/wiki/Path_dependence). Nonetheless, if clearer explanation can help us reach consensus more easily, it is worth writing more words.

Acknowledgement
---
Thanks all for you thoughtful feedback, including but not limited to u/FullTimeSadBoi, u/justinisrael and u/drvd. Please accept my apology if I have forgotten to mention your name.

**Comment by seankhliao** [comment]
I don't think we need a mirror of a reddit discussion in the issue tracker.
